### ISSUE TITLE
Generated uriPatterns create too many namespaces..

### ISSUE THREAD

**Issue by jgeluk** [d2rq/d2rq#277]
In a URI, the namespace part ends with the last slash or hash right?
So creating d2rq:uriPatterns with values like "column/@@id@@" generates a URI with a new namespace that ends with "column/" right? That's not nice. 

We should not have any slashes or hashes in uriPatterns.

**Comment by cygri** [comment]
Namespaces in RDF are a strange beast. There are multiple views that one could take.

On the one hand, one could say that they are purely a syntactic convenience for shortening repetitive URIs, and carry no syntactic meaning.

On the other hand, one could say that they are an integral part of the “user interface” of any RDF data, as users much prefer to interact with the data in its prefix-abbreviated form and not through the full URIs (especially in SPARQL queries).

Different parts of the RDF stack embody these different views. For example, the N-Triples syntax doesnt support namespace abbreviation at all. But the RDF/XML syntax doesn't work without namespace abbreviation (and furthermore makes it impossible to abbreviate certain valid URIs, but that's a different story).

> In a URI, the namespace part ends with the last slash or hash right?

It's sloppy to say that URIs have “namespace parts” and “local parts”. Given a certain prefix mapping, a URI can be broken up into a namespace part and a local part, possibly in different ways. Consider this:

```
@prefix p1: <http://example.com/>.
@prefix p2: <http://example.com/foo/>.
@prefix p3: <http://example.com/foo/bar->.
```

Now, given this prefix mapping, the URI `<http://example.com/foo/bar-123>` can be broken up into namespace and local parts in three different ways, yielding three different local names:

```
p1:foo/bar-123
p2:bar-123
p3:123
```

Now a couple of observations:
1. There's nothing inherently special about hashes or slashes in URI patterns, as shown in the third prefix.
2. It's a common convention to define prefix mappings that go up to the last hash or slash (again due to the influence of old RDF/XML where you couldn't have hashes or slashes in local names), and many tools that automatically create prefix mappings will do this, so you will see `p2:bar-123` more often than the other forms. But that's merely a convention, and the other forms may well be more convenient for users sometimes.
3. The form `p1:foo/bar-123` actually needs to be escaped as `p1:foo\/bar-123` if written in Turtle or SPARQL, because unescaped slashes are not allowed in the local part of a prefixed name.
4. This escape mechanism was only introduced in the W3C Recommendation version of Turtle and in SPARQL 1.1, so may not work in older parsers, and community awareness of this form is regrettably low.
5. The form `p3:123` will work in Turtle and SPARQL but not in RDF/XML, because XML requires that local names start with a character.

To summarise, URIs don't simply have a namespace part and local part. Rather, someone defines a prefix mapping, and under that prefix mapping, there may be zero, one or more ways of abbreviating any given URI.

Your argument seems to be that certain D2RQ-generated URIs lead to inconvenient results under a prefix mapping that is being automatically generated by some tool? Which tool? D2RQ itself? Where do you see those bad prefixes being applied?

Also, what change exactly would you like to see? Are you arguing that D2RQ should outlaw hashes or slashes in URI patterns, or warn users about them, or are you just seeking URI design advice?

**Comment by cygri** [comment]
I've [reposted](http://richard.cyganiak.de/blog/2016/02/uris-have-a-namespace-part-right/) my comment above on my blog, hope you don't mind.
